Assignment2Q1.java

```diff
@@ -0,0 +1,117 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class Assignment2Q1 {
+	public static void main(String[] args) {
+		BinaryTreeNode<Integer> n1 = new BinaryTreeNode<>(1, null, null);
+		BinaryTreeNode<Integer> n2 = new BinaryTreeNode<>(5, null, null);
+		BinaryTreeNode<Integer> n3 = new BinaryTreeNode<>(3, n1, n2);
+		BinaryTreeNode<Integer> n4 = new BinaryTreeNode<>(14, null, null);
+		BinaryTreeNode<Integer> n5 = new BinaryTreeNode<>(9, n3, n4);
+		BinaryTreeNode<Integer> n6 = new BinaryTreeNode<>(19, null, null);
+		BinaryTreeNode<Integer> n7 = new BinaryTreeNode<>(18, null, n6);
+		BinaryTreeNode<Integer> n8 = new BinaryTreeNode<>(16, n5, n7);
+
+		printAncestors(5, n8);
+	}
+
+	/**
+	 * Prints the ancestors of a node in a binary tree
+	 * 
+	 * @param key
+	 *            key of the node whose ancestors are going to be printed
+	 * @param root
+	 *            root of the binary tree
+	 * @throws IllegalArgumentException
+	 *             if key is not present in the tree
+	 */
+	public static <T> void printAncestors(T key, BinaryTreeNode<T> root) throws IllegalArgumentException {
+		List<T> list = findAncestors(key, root);
+
+		if (list == null) {
+			throw new IllegalArgumentException();
+		}
+
+		for (T el : list) {
+			System.out.print(el + " ");
+		}
+	}
+
+	/**
+	 * Helper method of printAncestors
+	 * 
+	 * @param key
+	 *            key of the node whose ancestors are going to be printed
+	 * @param root
+	 *            root of the binary tree
+	 * @return list of ancestors of the node
+	 */
+	private static <T> List<T> findAncestors(T key, BinaryTreeNode<T> root) {

```

Comment:
Small design comment: both findAncestors and findCommonAncestors operate on tree nodes and access their internal data (left, right etc). It is generally better to keep data and methods that operate on them together, so that it is clear that they depend on each other and to keep possible future changes in one place.

Would it make sense to move such functions and change them to (non-static) methods of BinaryTreeNode?